miden-crypto/src/merkle/forest/mod.rs

```diff
@@ -0,0 +1,176 @@
+
+    /// Leaves of all SMTs stored in this forest
+    leaves: BTreeMap<Word, SmtLeaf>,
+}
+
+impl Default for SmtForest {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl SmtForest {
+    // CONSTRUCTORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Creates an empty `SmtForest` instance.
+    pub fn new() -> SmtForest {
+        let store = MerkleStore::new();
+        let mut roots = BTreeSet::new();
+        let empty_tree_root = *EmptySubtreeRoots::entry(SMT_DEPTH, 0);
+        roots.insert(empty_tree_root);
+        let leaves: BTreeMap<Word, SmtLeaf> = BTreeMap::new();
+        SmtForest { store, roots, leaves }
+    }
+
+    // DATA EXTRACTORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns an opening for the specified key in the SMT with the specified root.
+    ///
+    /// Returns an error if an SMT with this root is not in the forest, or if the forest does
+    /// not have sufficient data to provide an opening for the specified key.
+    pub fn open(&self, root: Word, key: Word) -> Result<SmtProof, MerkleError> {
+        if !self.roots.contains(&root) {
+            return Err(MerkleError::RootNotInStore(root));
+        }
+
+        let index = key[3].as_int();
+        let node_index = NodeIndex::new(SMT_DEPTH, index)?;
+        let path: SparseMerklePath = self.store.get_path(root, node_index)?.path.try_into()?;
+
+        let leaf = match self.leaves.get(&key) {
+            Some(leaf) => leaf.clone(),
+            None => return Err(MerkleError::UntrackedKey(key)),
+        };
+
+        Ok(match SmtProof::new(path, leaf) {
+            Ok(proof) => proof,
+            Err(err) => match err {
+                SmtProofError::InvalidMerklePathLength(depth) => {
+                    return Err(MerkleError::InvalidPathLength(depth));
+                },
+            },
+        })
+    }
+
+    // STATE MUTATORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Inserts the specified key-value pair into an SMT with the specified root. This would also
+    /// add a new root to the forest.
+    ///
+    /// Returns an error if an SMT with the specified root is not in the forest, these is not
+    /// enough data in the forest to perform the insert, or if the insert would create a leaf
+    /// with too many entries.
+    pub fn insert(&mut self, root: Word, key: Word, value: Word) -> Result<Word, MerkleError> {
+        if !self.roots.contains(&root) {
+            return Err(MerkleError::RootNotInStore(root));
+        }
+
+        let index = key[3].as_int();
+        let node_index = NodeIndex::new(SMT_DEPTH, index)?;
+        let path = self.store.get_path(root, node_index)?.path;
+
+        let leaf_hash = match self.leaves.get_mut(&key) {
+            Some(leaf) => match leaf.insert(key, value) {
```

Comment:
Might be worth commenting that the `leaf` in `self.leaves` is mutated on `insert`. Took me a while to recognize that you update the set correctly.